drivers/hsc/hsc.c

```
@@ -0,0 +1,151 @@
+ *              temperature and pressure sensor.
+ *
+ * @author      Quang Pham <phhr_quang@live.com>
+ *
+ * @}
+ */
+
+#include "hsc.h"
+#include "errno.h"
+#include "hsc_internals.h"
+#include "hsc_params.h"
+#include "periph/i2c.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+#define DEV_I2C      (dev->params.i2c_dev)
+#define DEV_ADDR     (dev->params.i2c_addr)
+
+/* Internal function prototypes */
+static int _read_ut(const hsc_t *dev, int32_t *ut);
+static int _read_up(const hsc_t *dev, int32_t *up);
+
+/*---------------------------------------------------------------------------*
+ *                          HSC Core API                                     *
+ *---------------------------------------------------------------------------*/
+
+int hsc_init(hsc_t *dev, const hsc_params_t *params)
+{
+    dev->params = *params;
+    uint8_t buf[HSC_FULL_DATA_LENGTH];
+
+    /* Acquire exclusive access */
+    if (i2c_acquire(DEV_I2C)) {
+        return -ENODEV;
+    }
+
+    if (i2c_read_bytes(DEV_I2C, DEV_ADDR, buf, sizeof(buf), 0) < 0) {
+        i2c_release(DEV_I2C);
+        DEBUG("[hsc] read_bytes fails\n");
+
+        return -EIO;
+    }
+
+    i2c_release(DEV_I2C);
+
+    /* Check the status of the sensor */
+    uint8_t status = buf[0];
+
+    switch (status & HSC_STATUS_MASK) {
+    case HSC_STATUS_OK:
+        return 0;
+    case HSC_STATUS_STALE_DATA:
+        return -EAGAIN;
+    default:
+        return -EIO;
+    }
+
+    return 0;
+}
+
+int16_t hsc_read_temperature(const hsc_t *dev)
+{
+    int32_t ut = 0;
+
+    /* Read uncompensated temperature value */
+    _read_ut(dev, &ut);
+
+    /* Formular from section 4.0 in 
+     * https://sensing.honeywell.com/i2c-comms-digital-output-pressure-sensors-tn-008201-3-en-final-30may12.pdf */
+    return (int16_t)(ut * 2000 / 2047 - 500);
+}
+
+int32_t hsc_read_pressure(const hsc_t *dev)
+{
+    int32_t up = 0;
+
+    /* Read uncompensated pressure value */
+    _read_up(dev, &up);
```

Here again, potential errors are not handled. However, as pressure cannot be negative, you can just return the error codes here. So any non-negative value would indicate that the return value is indeed pressure information, every negative value would indicate an error instead.

---

the pressure can be negative too, so I modified the function together with the temperature.